service/src/main/java/org/ehrbase/service/DirectoryServiceImp.java

```diff
@@ -163,16 +166,29 @@ public Folder update(UUID ehrId, Folder folder, ObjectVersionId ifMatches) {
 
     @Override
     public Folder update(UUID ehrId, Folder folder, ObjectVersionId ifMatches, UUID contributionId, UUID auditId) {
+
+        UUID uuid = UUID.fromString(ifMatches.getObjectId().getValue());
+
+        // Ensure UIDs matching
+        String folderUidValue = Optional.ofNullable(folder.getUid())
+                .map(UIDBasedId::getRoot)
+                .map(UID::getValue)
+                .orElse(null);
+
+        if (folderUidValue != null && !Objects.equals(uuid.toString(), folderUidValue)) {
+            throw new PreconditionFailedException(
+                    String.format("FOLDER uid %s does not match %s", folderUidValue, uuid));
+        }
+
         // validation
         ehrService.checkEhrExistsAndIsModifiable(ehrId);
-        if (!ehrFolderRepository.hasFolder(ehrId, EHR_DIRECTORY_FOLDER_IDX)) {
+        if (!ehrFolderRepository.hasFolderForVoId(uuid)) {
             throw new PreconditionFailedException(
                     String.format("EHR with id %s does not contain a directory.", ehrId.toString()));
```

Comment:
We should have different messages for uid mismatch and ehr without directory

Reply:
Changed, now running both checks #1 for has folder with idx and a second for folder with voId.Downside is that new 2 round trips are performed.

Reply:
You could either run ::hasFolderInEhrForVoId first and ::hasFolderAtIndex only for the message, or let ::hasFolderAtIndex return the vo_id and check it in java

Reply:
changed to use first `hasFolderInEhrForVoId` and `hasFolderAtIndex` only for a more precise error message.